### ISSUE TITLE
Nami wallet does not take into account funds at enterprise address

### ISSUE THREAD

**Issue by MuesliSwapTeam** [Berry-Pool/nami-wallet#175]
For a given public key hash, Nami wallet displays all funds locked at the primary wallet address. In case the wallet is imported in other wallets, it also takes into account other wallet addresses and correctly displays that funds there are not locked at the primary address. 

However, Nami does not take into account funds locked at the controlled enterprise address (denoted in HEX by "61{pubkeyhash}", without a staking key). This might lead to confusion among users that received funds on these addresses. Wallets like CCVault and Yoroi correctly display funds at these addresses. It would be favorable of Nami to also take funds at these addresses into account.

**Comment by MuesliSwapTeam** [comment]
Some example data:
This is the primary address created by a Nami wallet used in our team:
[addr1qy236l8zxnyvvm7uaxtvp87k36a045h8em0y82clams4wfuwpj9clsvsf85cd4xc59zjztr5zwpummwckmzr2myjwjnsrv508n](https://cardanoscan.io/address/addr1qy236l8zxnyvvm7uaxtvp87k36a045h8em0y82clams4wfuwpj9clsvsf85cd4xc59zjztr5zwpummwckmzr2myjwjnsrv508n)

Any dApp or third party that decides to send money to an address with the same public key hash, but does not know staking information or other information could derive this enterprise address:
[addr1vy236l8zxnyvvm7uaxtvp87k36a045h8em0y82clams4wfc3u5sce](https://cardanoscan.io/address/61151d7ce234c8c66fdce996c09fd68ebafad2e7cede43ab1feee15727)

While the funds are completely safe and controlled by the same user as the Nami Wallet central address as they are based upon the same public key hash, the funds are not shown inside Nami Wallet.

**Comment by MuesliSwapTeam** [comment]
See https://cips.cardano.org/cips/cip19/#shelleyaddresses for more details on Shelley type addresses

**Comment by Jeyhey** [comment]
Many of the examples of the Plutus repository show how to pay funds to a PubKeyHash-address (enterprise address). Only the newest Nami-demo shows how to take into account the staking key hash. I suppose that many dApps will miss this subtle difference and try to pay to the users enterprise address.

**Comment by alessandrokonrad** [comment]
Nami is a single address wallet. Adding more addresses adds complexity and destroys the ease of use. Also I assume a user would not even like to have the smart contract return the balance to a non staking address.
It's the dApps task to handle this, the wallet provides you through the dApp connector all relevant information regarding what addresses can be used.

**Comment by Jeyhey** [comment]
I see your point. On the other hand, this means that the wallet does only show a subset of accessible funds. This will eventually lead to confusion and an impaired user experience. Users will get the impression their funds are lost when they are not. For instance the current MuesliSwap contracts settle the orders only using the PubKeyHash: [see-here](https://github.com/MuesliSwapTeam/muesliswap-cardano-contracts/blob/main/order-validator/src/Cardano/MuesliSwapOrderValidator/OrderValidator.hs)

**Comment by MuesliSwapTeam** [comment]
We are aware that Nami does already handle cases where funds are stored in another wallet address that contains staking keys. It would be nice to have a similar kind of information about funds stored at enterprise addresses.

![image](https://user-images.githubusercontent.com/91151317/147991883-efe4a95c-26bb-4910-a8ba-6c9c5202768c.png)

**Comment by alessandrokonrad** [comment]
> On the other hand, this means that the wallet does only show a subset of accessible funds.

I disagree. The PubKeyHash is not an address. If you argue like that the wallet should be able to find next to enterprise addresses, also all pointer addresses including this PubKeyHash, which is pretty much impossible.

> For instance the current MuesliSwap contracts settle the orders only using the PubKeyHash: see-here

Why not replace PubKeyHash with Address here?

**Comment by MuesliSwapTeam** [comment]
> Why not replace PubKeyHash with Address here?

This is possible but requires a contract upgrade, which will take time. You are aware of the fact that a lot of live Smart Contracts validate received funds only by the PubKeyHash, for example [SpaceBudz](https://github.com/Berry-Pool/spacebudz/blob/main/src/cardano/market/contract.hs#L147) and [MartifyLabs](https://github.com/MartifyLabs/martify.contracts/blob/main/src/Market/Onchain.hs#L78)?

**Comment by MuesliSwapTeam** [comment]
> If you argue like that the wallet should be able to find next to enterprise addresses, also all pointer addresses including this PubKeyHash, which is pretty much impossible.

It is however certainly possible to find the enterprise address, and in our opinion for that sole reason users should be informed about the fact that they own funds at these addresses. Further as this is already done for other types of addresses, it should not cause much trouble.

**Comment by alessandrokonrad** [comment]
> > Why not replace PubKeyHash with Address here?
> 
> This is possible but requires a contract upgrade, which will take time. You are aware of the fact that a lot of live Smart Contracts validate received funds only by the PubKeyHash, for example [SpaceBudz](https://github.com/Berry-Pool/spacebudz/blob/main/src/cardano/market/contract.hs#L147) and [MartifyLabs](https://github.com/MartifyLabs/martify.contracts/blob/main/src/Market/Onchain.hs#L78)?

I am aware and I don't say it's perfect. That is why I am for a change. PubKeyHash alone is not sufficient. Someone could make a trade and pay to someone's payment key hash, but with the stake key of his choice. Basically an address that is first of all hard to retrieve and secondly the receiver does certainly not like. The examples of IOG are kinda misleading

**Comment by MuesliSwapTeam** [comment]
Totally agreed that this should change in the future! However, we will see many more users that will end up with funds on enterprise addresses as more dApps emerge. On MuesliSwap, users can already reclaim funds on the enterprise address with a simple click. Other dApps will hopefully implement similar features if their SCs are only based on PubKeyHashes, and in the long term fix their SCs.

However it should be in the interest of Nami to prevent users from panicking over lost funds. Otherwise dApps might simply start recommending the usage of other wallets that correctly display owned funds in the interest of their users.
